src/server/scripts/EasternKingdoms/ScarletMonastery/boss_mograine_and_whitemane.cpp

```diff
@@ -61,321 +53,391 @@ enum Spells
-        }
+    /// Mograine
+    EVENT_CRUSADER_STRIKE = 2,
+    EVENT_HAMMER_OF_JUSTICE,
 
-        void Initialize()
-        {
-            CrusaderStrike_Timer = 10000;
-            HammerOfJustice_Timer = 10000;
-            _bHasDied = false;
-            _bHeal = false;
-            _bFakeDeath = false;
-        }
+    /// Whitemane
+    EVENT_HEAL = 2,
+    EVENT_POWER_WORD_SHIELD,
+    EVENT_HOLY_SMITE,
+};
 
-        InstanceScript* instance;
+enum Points
+{
+    POINT_WHITEMANE_MOVE_TO_MOGRAINE = 1,
+};
 
-        uint32 CrusaderStrike_Timer;
-        uint32 HammerOfJustice_Timer;
+const Position WhitemaneIntroMovePos(1163.113370f, 1398.856812f, 32.527786f);
 
-        bool _bHasDied;
-        bool _bHeal;
-        bool _bFakeDeath;
+/// Scarlet Commander Mograine - 3976
+struct boss_scarlet_commander_mograine : public ScriptedAI
+{
+    public:
+        boss_scarlet_commander_mograine(Creature* creature) : ScriptedAI(creature)
+        {
+            m_CanTalkKill = true;
+            m_CanDie = false;
+            m_FakeDeath = false;
+        }
 
         void Reset() override
         {
-            Initialize();
+            m_CanTalkKill = true;
+            m_CanDie      = false;
+            m_FakeDeath   = false;
 
-            //Incase wipe during phase that mograine fake death
-            me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE);
+            events.Reset();
+            scheduler.CancelAll();
+
+            DoCast(me, SPELL_RETRIBUTIONAURA);
             me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE);
             me->SetStandState(UNIT_STAND_STATE_STAND);
+            me->SetReactState(REACT_AGGRESSIVE);
 
-            if (me->IsAlive())
-                instance->SetBossState(DATA_MOGRAINE_AND_WHITE_EVENT, NOT_STARTED);
-        }
-
-        void JustReachedHome() override
-        {
-            if (instance->GetBossState(DATA_MOGRAINE_AND_WHITE_EVENT) != NOT_STARTED)
-                instance->SetBossState(DATA_MOGRAINE_AND_WHITE_EVENT, FAIL);
+            // Close the door
+            if (auto instance = me->GetInstanceScript())
+                instance->HandleGameObject(instance->GetGuidData(GO_HIGH_INQUISITORS_DOOR), false);
         }
 
         void JustEngagedWith(Unit* /*who*/) override
         {
             Talk(SAY_MO_AGGRO);
-            DoCast(me, SPELL_RETRIBUTIONAURA);
 
+            // Call for help (from old script)
             me->CallForHelp(VISIBLE_RANGE);
+
+            // Schedule events
+            events.ScheduleEvent(EVENT_CRUSADER_STRIKE, Seconds(10), Seconds(15));
+            events.ScheduleEvent(EVENT_HAMMER_OF_JUSTICE, Seconds(10), Seconds(15));
         }
 
-        void KilledUnit(Unit* /*victim*/) override
+        void KilledUnit(Unit* who) override
         {
+            if (who->GetTypeId() != TYPEID_PLAYER || !m_CanTalkKill)
+                return;
+
+            // Prevent talking more than once - resets every 5 seconds by EVENT_RESET_KILL_TALK
```

Comment:
what about using a single time_t variable instead ? would that look less complicated than a flag and an event resetting it ?